### PR TITLE
Improve edge case coverage of load_node magic

### LINKED ISSUE #5417: Improve %load_node to cover edge case when functions are defined in same module
## Description

With the default spaceflight project, `%load_node` doesn't work on all types of nodes. `%load_node` currently surface the necessary definition of a node by tracing necessary imports.

- It parses all the import statement (not function specific)
- function body including the signature

While this works well for most case, it can still missed some cases (like the one in spaceflights), where there are some other (helper) function defined in the same module, but lives outside of the function. See examples

To load a node in notebook cell:
```
%load_node preprocess_companies_node
```

This will then appears in next cell

```python
# Prepare necessary inputs for debugging
# All debugging inputs must be defined in your project catalog
companies = catalog.load("companies")

import pandas as pd

############This is not included currently###########################
def _is_true(x: pd.Series) -> pd.Series:
    return x == "t"

def _parse_percentage(x: pd.Series) -> pd.Series:
    x = x.str.replace("%", "")
    x = x.astype(float) / 100
    return x
###############################


def preprocess_companies(companies: pd.DataFrame) -> pd.DataFrame:
    """Preprocesses the data for companies.

    Args:
        companies: Raw data.
    Returns:
        Preprocessed data, with `company_rating` converted to a float and
        `iata_approved` converted to boolean.
    """
    companies["iata_approved"] = _is_true(companies["iata_approved"])
    companies["company_rating"] = _parse_percentage(companies["company_rating"])
    return companies


preprocess_companies(companies)
```

As a result, this wouldn't executed in notebook becaue `_parse_percent` , `_is_true` are missing symbols.

## Context

This breaks the UX to provide a smooth debugging experience.
## Possible Implementation

AST parsing as a best-effort solution. PR will follow.

### PR DESCRIPTION
**Pull request by noklam**
## Description

Fix #5417,  there are details what this PR try to fix. In addition, the test case should be self-explanatory `test_load_node_executes_with_same_module_helper_dependency`

The implementation includes a fallback strategy, in case there are rare situation that source code is not reachable.

Related: https://github.com/kedro-org/vscode-kedro/pull/257

Some of the nodes in spaceflights use helper function that are defined in the same module. Without this fix, %load_node will break on these edge cases.

## Developer Certificate of Origin
We need all contributions to comply with the [Developer Certificate of Origin (DCO)](https://developercertificate.org/). All commits must be signed off by including a `Signed-off-by` line in the commit message. [See our wiki for guidance](https://github.com/kedro-org/kedro/wiki/Guidelines-for-contributing-developers/).

If your PR is blocked due to unsigned commits, then you must follow the instructions under "Rebase the branch" on the GitHub Checks page for your PR. This will retroactively add the sign-off to all unsigned commits and allow the DCO check to pass.

## Checklist

- [ ] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [ ] Signed off each commit with a [Developer Certificate of Origin (DCO)](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-the-commit-signoff-policy-for-your-repository)
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes
- [ ] Checked if this change will affect Kedro-Viz, and if so, communicated that with the Viz team

### CODE REVIEW

**Comment by ravi-kumar-pilla** [comment]
Hi @noklam ,

Can you please resolve the conflicts and also unit tests ? Any QA notes would help test this quickly. Thank you

**Comment by noklam** [comment]
test_load_node_executes_with_same_module_helper_dependency

This is the main test that cover the edge case. Sorry I don't have access to computer.

**Comment by jitu5** [comment]
Thanks @noklam, nice improvement overall. One issue before merge: Decorators are silently stripped

The new path uses `ast.get_source_segment(module_source, node)`, which starts at `FunctionDef.lineno` (the `def` line) and excludes the decorator list. The previous `inspect.getsourcelines()` path preserved decorators.
  Example, a helper like:

```
  @functools.lru_cache
  def _is_true(x):
      return x == "t"
```

will appear in the generated cell as:

```
  def _is_true(x):
      return x == "t"
```

**Comment by noklam** [comment]
Thanks. So we need this test case. I will revise the PR.

**Comment by noklam** [comment]
Fixed - it is indeed a regression as the implementation change from `inspect` to `ast`. I have an additional test case to cover this.

**Review by jitu5 — APPROVED** [review]

File: kedro/ipython/__init__.py [hunk 1]
```diff
@@ -423,9 +424,195 @@ def _format_node_inputs_text(input_params_dict: dict[str, str] | None) -> str |
...
+def _resolve_function_node(
+    module_tree: ast.Module, func: Callable
+) -> ast.FunctionDef | ast.AsyncFunctionDef:
+    """Resolve the top-level AST node that corresponds to ``func``."""
+    _, source_line = inspect.getsourcelines(func)
+    candidates = [
+        node
+        for node in module_tree.body
+        if isinstance(node, ast.FunctionDef | ast.AsyncFunctionDef)
+        and node.name == func.__name__
```
Comment (ravi-kumar-pilla) [inline kedro/ipython/__init__.py:537]: Does this work with nested functions too ?

**Review by ravi-kumar-pilla — APPROVED** [review]
Left a minor comment but otherwise works well. Code seems a bit complex to follow and might be due to the whole AST parsing and mapping which is needed to address the issue. Thank you @noklam

**Review by rashidakanchwala — APPROVED** [review]
Hi @noklam, thanks for this. The implementation is complex, but I think it's a good solution with appropriate fallback behavior. Please fix any remaining CI issues and we can merge.

**Review by rashidakanchwala — APPROVED** [review]

File: kedro/ipython/__init__.py (hunk 1 above)
Comment (noklam) [reply kedro/ipython/__init__.py:537]: Not in the most generic sense.

```
def dummy_nested_function(dummy_input):
    def nested_function(input):
        return not input

    return nested_function(dummy_input)
```

This should work
